src/Nethermind/Nethermind.Baseline/MerkleTree.cs

```diff
@@ -0,0 +1,232 @@
+            return indexAtLevel;
+        }
+
+        internal static uint GetLeafIndex(uint nodeIndex)
+        {
+            if (nodeIndex == uint.MaxValue ||
+                nodeIndex < FirstLeafIndexAsNodeIndex)
+            {
+                throw new IndexOutOfRangeException($"Leaf indices start at {FirstLeafIndexAsNodeIndex}");
+            }
+
+            return nodeIndex - FirstLeafIndexAsNodeIndex;
+        }
+
+        internal static uint GetSiblingIndex(uint level, uint indexAtLevel)
+        {
+            ValidateLevel(level);
+            ValidateIndexAtLevel(level, indexAtLevel);
+
+            if (level == 0)
+            {
+                throw new IndexOutOfRangeException("Root node has no siblings.");
+            }
+            
+            if (indexAtLevel % 2 == 0)
+            {
+                return indexAtLevel + 1;
+            }
+
+            return indexAtLevel - 1;
+        }
+
+        internal static void ValidateNodeIndex(uint nodeIndex)
+        {
+            if (nodeIndex > MaxNodeIndex)
+            {
+                throw new IndexOutOfRangeException($"Node index should be between 0 and {MaxNodeIndex}");
+            }
+        }
+
+        internal static void ValidateLevel(uint level)
+        {
+            if (level > LeafLevel)
+            {
+                throw new IndexOutOfRangeException($"Tree level should be between 0 and {LeafLevel}");
+            }
+        }
+
+        internal static void ValidateIndexAtLevel(uint level, uint indexAtLevel)
+        {
+            uint maxIndexAtLevel = (1u << (int) level) - 1u;
+            if (indexAtLevel > maxIndexAtLevel)
+            {
+                throw new IndexOutOfRangeException($"Tree level {level} should only has indices between 0 and {maxIndexAtLevel}");
+            }
+        }
+
+        internal static uint GetNodeIndex(uint level, uint indexAtLevel)
+        {
+            ValidateLevel(level);
+            ValidateIndexAtLevel(level, indexAtLevel);
+
+            return (1u << (int) level) - 1u + indexAtLevel;
+        }
+
+        internal static uint GetParentIndex(uint nodeIndex)
+        {
+            ValidateNodeIndex(nodeIndex);
+
+            if (nodeIndex == 0)
+            {
+                throw new IndexOutOfRangeException("Root node has no parent");
+            }
+
+            return (nodeIndex + 1) / 2 - 1;
+        }
+
+        [MethodImpl(MethodImplOptions.Synchronized)]
+        public void Insert(Bytes32 leaf)
+        {
+            _keyValueStore[Rlp.Encode(Count).Bytes] = leaf.AsSpan().ToArray();
+
+            uint indexAtLevel = Count;
+            uint siblingIndexAtLevel = GetSiblingIndex(LeafLevel, indexAtLevel);
+            Bytes32 hash = leaf;
+            Bytes32 siblingHash = LoadValue(LeafLevel, siblingIndexAtLevel);
+
+            uint nodeIndex = GetNodeIndex(LeafLevel, Count);
+            SaveValue(nodeIndex, hash);
+
+            for (int level = LeafLevel; level > 0; level--)
```

Comment:
make level uint

Reply:
risky to use uint because if you change > 0 to >=0 then you may end up with an infinite loop